### PR TITLE
Per segment AO versions

### PR DESCRIPTION
**Pull request by hlinnaka**
This patch series contains a bunch of Append-only and AOCS-related changes. I ran into these while working on pg_upgrade for GPDB.

The crux of these patches is the second to last patch, which turns the "file format version" number from a per-table property, to a per-segment property. That way, if you have an AO table that's in the old version, and you upgrade, we don't necessarily need to rewrite the whole table at one go, to change the format. As long as we can read the old format, we can keep the existing segments in the old format, and direct any new data to new segments, in new format. That will come handy with pg_upgrade.

### CODE REVIEW

**Comment by hlinnaka** [comment]
@asimrp, please have a look. We had a discussion about this off-list. I'm not sure if I was able to explain how this would work, so please don't hesitate to ask, if this looks like nonsense.

**Comment by hlinnaka** [comment]
For more context, there's a Work-In-Progress branch to backport pg_upgrade, for upgrading from 4.3 to 5.0, at https://github.com/hlinnaka/gpdb/tree/pg_upgrade. I just rebased it over this PR. In particular, see https://github.com/hlinnaka/gpdb/commit/3fe7d97bb1606aa7739c12c85862c5a98945dd33, which adds the code to convert numerics in AO tables from old format to new.

File: src/backend/access/appendonly/aosegfiles.c [hunk 1]
```diff
@@ -233,6 +246,14 @@ GetFileSegInfo(Relation parentrel, AppendOnlyEntry *aoEntry, Snapshot appendOnly
 				(errcode(ERRCODE_UNDEFINED_OBJECT),
 				errmsg("got invalid modcount value: NULL")));
 
+	/* get the file format version number */
+	formatversion = fastgetattr(tuple, Anum_pg_aoseg_formatversion, pg_aoseg_dsc, &isNull);
+
+	if(isNull)
```
Comment (danielgustafsson) [inline src/backend/access/appendonly/aosegfiles.c]: nitpick: spacing between `if` and `(`?

**Comment by danielgustafsson** [comment]
Reading over this it seems very reasonable, I like the concept and the implementation. I think however we could skip the passing the version parameter to (and the error handling of it) `AppendOnlyStorageWrite_OpenFile()`, I don't see why we'd ever relax the requirement to write using the latest version?

Admittedly, my eyes did glaze over on ec6e50cb96b2a78ec3e27a29bc24f0ea889b8e9c so I might've missed something there but it seemed quite straight forward (albeit extremely useful, thanks for tackling this!)

**Comment by hlinnaka** [comment]
I pushed the refactoring and cleanup commits, and rebased the two substantial ones. @asimrp, can you also have a look when you have a chance, please?

**Comment by asimrp** [comment]
Hey Heikki, sorry for no response at all.  I'm going to look at it today.

File: src/backend/access/aocs/aocsam.c [hunk 2]
```diff
@@ -101,7 +102,7 @@ static void open_all_datumstreamread_segfiles(Relation rel,
     for(i=0; i<nvp; ++i)
     {
         if (proj[i])
-	{
+		{
```
Comment (asimrp) [inline src/backend/access/aocs/aocsam.c]: cosmetic: the block from line 105 - 117 should be indented one tab to the left.

File: src/backend/access/appendonly/aosegfiles.c [hunk 3]
```diff
@@ -230,6 +243,14 @@ GetFileSegInfo(Relation parentrel, Snapshot appendOnlyMetaDataSnapshot, int segn
...
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				errmsg("got invalid state value: NULL")));
```
Comment (asimrp) [inline src/backend/access/appendonly/aosegfiles.c]: "got invalid formatversion value"

File: src/backend/access/aocs/aocssegfiles.c [hunk 4]
```diff
@@ -87,6 +100,7 @@ InsertInitialAOCSFileSegInfo(Relation prel, int4 segno, int4 nvp)
     values[Anum_pg_aocs_vpinfo-1] = PointerGetDatum(vpinfo);
 	values[Anum_pg_aocs_tupcount-1] = Int64GetDatum(0);
 	values[Anum_pg_aocs_varblockcount-1] = Int64GetDatum(0);
+	values[Anum_pg_aocs_formatversion-1] = Int16GetDatum(formatVersion);
```
Comment (asimrp) [inline src/backend/access/aocs/aocssegfiles.c]: We are using one version column in pg_aocsseg relation to record version of a aocs segment.  However, `datumstreamwrite_open_file()` is modified above to accept version as an argument.  Which tells me that we want enable different columns' segfiles to have different versions.  There seems to be a discrepancy here, unless I'm missing something.

File: src/backend/access/appendonly/appendonlyam.c [hunk 5]
```diff
@@ -985,36 +1052,22 @@ AppendOnlyExecutorReadBlock_ProcessTuple(
...
-			ExecStoreMinimalTuple(tuple, slot, true /* shouldFree */);
-		}
-		else
-		{
-			ExecStoreMinimalTuple(tuple, slot, false);
-		}
+		bool		shouldFree = false;
 
+		/* If the tuple is not in the latest format, convert it */
+		if (formatVersion < AORelationVersion_GetLatest())
```
Comment (asimrp) [inline src/backend/access/appendonly/appendonlyam.c]: Just to confirm, this condition will be true for the AO tables that had a numeric column and the column existed before upgrade.

Comment (asimrp) [inline src/backend/access/appendonly/appendonlyam.c]: Never mind, got the answer after looking at your upgrade branch.  The latest version returned by `AORelationVersion_GetLatest()` will be  `AORelationVersion_PG83`.  That means, for every AO tuple from a table that existed prior to upgrade, `upgrade_tuple()` will be invoked.  Did I understand that correctly?

Comment (hlinnaka) [inline src/backend/access/appendonly/appendonlyam.c]: On 08/26/2016 04:03 AM, Asim RP wrote:
>>
>> +		/* If the tuple is not in the latest format, convert it */
>> +		if (formatVersion < AORelationVersion_GetLatest())
>
> Never mind, got the answer after looking at your upgrade branch.  The
> latest version returned by `AORelationVersion_GetLatest()` will be
> `AORelationVersion_PG83`.  That means, for every AO tuple from a
> table that existed prior to upgrade, `upgrade_tuple()` will be
> invoked.  Did I understand that correctly?

Yep!

- Heikki

File: src/backend/access/aocs/aocssegfiles.c (hunk 4 above)
Comment (hlinnaka) [inline src/backend/access/aocs/aocssegfiles.c]: On 08/26/2016 03:17 AM, Asim RP wrote:
>> @@ -87,6 +100,7 @@ InsertInitialAOCSFileSegInfo(Relation prel, int4 segno, int4 nvp)
>>      values[Anum_pg_aocs_vpinfo-1] = PointerGetDatum(vpinfo);
>>  	values[Anum_pg_aocs_tupcount-1] = Int64GetDatum(0);
>>  	values[Anum_pg_aocs_varblockcount-1] = Int64GetDatum(0);
>> +	values[Anum_pg_aocs_formatversion-1] = Int16GetDatum(formatVersion);
>
> We are using one version column in pg_aocsseg relation to record
> version of a aocs segment.  However, `datumstreamwrite_open_file()`
> is modified above to accept version as an argument.

We need to pass the version to datumstreamwrite_open_file() in either 
case, because it needs to know what version to write in (after the 
second commit in the series, we could assume that we always write in the 
latest format, though). We used to pass it in create_datumstreamwrite(), 
but now that the version is a per-segment property, it can change when 
you switch from one segment to another, with datumstreamwrite_open_file().

In practice, I don't think we ever use a datumstreamwriter in that way. 
We only call datumstreamwrite_open_file() once, after a 
create_datumstreamwrite() call, and we don't switch segments when 
writing. Reading is a different story. But since the version number is a 
per-segment property, passing it in the datumstreamwrite_open_file() 
call is conceptually the right place.

> Which tells me that we want enable different columns' segfiles to
> have different versions.

Not at the moment. But we could, and it might be a useful thing to do in 
the future. We don't do any write operations on individual column 
segments, so it wouldn't make any difference at the moment. But if we 
could e.g. vacuum only a single column segfile, then it would be useful, 
and necessary, to have a different version number for each column's segfile.

- Heikki

**Comment by asimrp** [comment]
The change looks good, overall.  A couple of points:
1. I'm not familiar with pg_upgrade but I'm hoping that the AO tables that do not need to be rewritten (i.e. not having a `numeric` column) are assigned the latest version number, so their tuples won't suffer performance penalty of `upgrade_tuple()`.
2. After upgrade, how is the PR preventing writes to segment files with old version?  My understanding is, old version segment files should remain read-only and at an opportune time, users are expected to run CTAS to rewrite the content to the latest version segment file.

File: src/backend/access/aocs/aocssegfiles.c (hunk 4 above)
Comment (asimrp) [inline src/backend/access/aocs/aocssegfiles.c]: Makes sense, thank you for the clarification.

**Comment by hlinnaka** [comment]
On 08/26/2016 09:28 AM, Asim RP wrote:
> The change looks good, overall.  A couple of points:
> 1. I'm not familiar with pg_upgrade but I'm hoping that the AO tables
> that do not need to be rewritten (i.e. not having a `numeric` column)
> are assigned the latest version number, so their tuples won't suffer
> performance penalty of `upgrade_tuple()`.

Good idea. The current WIP code we have doesn't do that, but it could.

> 2. After upgrade, how is the PR preventing writes to segment files
> with old version?

This change in appendonlywriter.c does it:

https://github.com/greenplum-db/gpdb/pull/995/files#diff-4ba54983f876862f1e1620a24a58d339L1145

So, when choosing a segment to write to, in addition to checking that a 
segment is not currently in use, we also check that it's in the latest 
format, and skip it if not.

Come to think of it, we have a problem with segfile #0 here. After 
upgrade, segfile #0 will be in the old format, so if you try to load 
more data in utility mode, it will fail, because in utility mode we 
always write to segfile #0. Hopefully we get 
https://github.com/greenplum-db/gpdb/pull/790 into GPDB 5, which will 
fix that problem.

Another potential problem is that if the table already had all 127 
segments in use before the upgrade, inserting after the upgrade will 
fail, because it cannot create a new segment in the new format.

Perhaps we should bump the max number of segments to avoid that problem. 
If we have to double the maximum on every major release, though, to deal 
with this same problem, it'll quickly get out of hand though. Then 
again, we're not anticipating any further format changes in the next few 
releases. There aren't further datatype format changes in the upstream 
that would force that.

> My understanding is, old version segment files should remain
> read-only and at an opportune time, users are expected to run CTAS to
> rewrite the content to the latest version segment file.

Basically, yes. A vacuum should be enough to rewrite it, no need for a 
CTAS which would rewrite all segments. We might need to teach vacuum to 
rewrite a segment that's in the old format, even when there hasn't been 
any changes to it, though. But that's not urgent. And there's no rush 
for a customer to rewrite the tables either, there's no user-visible 
difference between a segment a that's in the old or the new format.

- Heikki

**Comment by asimrp** [comment]
On Fri, Aug 26, 2016 at 12:07 AM, Heikki Linnakangas <
notifications@github.com> wrote:
>
> > 2. After upgrade, how is the PR preventing writes to segment files
> > with old version?
>
> This change in appendonlywriter.c does it:
>
>
https://github.com/greenplum-db/gpdb/pull/995/files#diff-4ba54983f876862f1e1620a24a58d339L1145
>
> So, when choosing a segment to write to, in addition to checking that a
> segment is not currently in use, we also check that it's in the latest
> format, and skip it if not.

Got it, thanks.

>
> Another potential problem is that if the table already had all 127
> segments in use before the upgrade, inserting after the upgrade will
> fail, because it cannot create a new segment in the new format.
>
> Perhaps we should bump the max number of segments to avoid that problem.
> If we have to double the maximum on every major release, though, to deal
> with this same problem, it'll quickly get out of hand though. Then
> again, we're not anticipating any further format changes in the next few
> releases. There aren't further datatype format changes in the upstream
> that would force that.

Another simpler alternative could be to explicitly warn about such tables
during upgrade and let DMLs fail post upgrade.  The users will have an
option to run vacuum (assuming we teach vacuum to rewrite old version
segfiles into latest ones) on such tables or to recreate them using CTAS.
I understand this sounds hand wavy, but it relieves us from complicating AO
segfile selection logic even further.

Asim

**Comment by danielgustafsson** [comment]
I'm fixing pg_upgrade to warn about segfile exhaustion with the CTAS option in the text in the first version before going on to fix vacuum (CTAS might be prohibitively expensive in some cases but one thing at a time). Is there anything else stopping this from going in?

File: src/backend/access/aocs/aocsam.c (hunk 2 above)
Comment (danielgustafsson) [inline src/backend/access/aocs/aocsam.c]: This is just a space vs tabs issue which seems strange in the Github rendering.

File: src/backend/access/appendonly/aosegfiles.c (hunk 3 above)
Comment (danielgustafsson) [inline src/backend/access/appendonly/aosegfiles.c]: Fixed in rebased branch.

**Comment by danielgustafsson** [comment]
Rebased this on top of current master and fixed the `errmsg()` comment by @asimrp

**Comment by danielgustafsson** [comment]
And again rebased over current master with fixes.

**Comment by danielgustafsson** [comment]
@asimrp @hlinnaka with Concourse and Travis green on this rebased and slightly fixed branch, are we ready to push this to master? Anything left stopping?

**Comment by danielgustafsson** [comment]
Rebased again, in case Concourse is green and happy with this I will push this to master.

**Comment by danielgustafsson** [comment]
Pushed to master.
